### ISSUE TITLE
Server runs on Windows not on Mac

### ISSUE THREAD

**Issue by tobudim** [OfficeDev/generator-office#570]
- [x] I am running the latest version of Node and the tools
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

# Expected behavior

The add-in server should run on Windows and Mac.

# Current behavior

The add-in server runs on Windows.
The add-in server fails on Mac.

## Steps to Reproduce

1. Generate an Outlook dedicated add-in with SSO.
2. _Note: I can't tell everything I did on the add-in though I'm using preview Outlook API with signatures and a call to a HTTPS server of mine._
3. Run the server with `npm start` on Windows 10: should work just fine.
4. Run the server on the latest Mac OS Big Sur with `npm start`: should not work, even in sudo.

## Context

Mac
* Operating System: Mac OS Big Sur v11.0.1
* Node version: 14.15.0 & 14.15.1
* NPM version: 6.14.8

Windows
* Operating System: Windows 10 v19042.630
* Node version: 14.15.1
* NPM version: 6.14.8

## Failure Logs

The terminal on Mac outputs this:

```bash
[0] > office-addin-taskpane-sso-js@0.0.0 start:server /Users/tobudim/code/w10test
[0] > office-addin-sso start manifest.xml
[0] 
[0] (node:2227) UnhandledPromiseRejectionWarning: false
[0] (Use `node --trace-warnings ...` to show where the warning was created)
[0] (node:2227) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
[0] (node:2227) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[0] npm run start:server exited with code 0
```

**Comment by tobudim** [comment]
Actually, I have the same issue on 2 other Windows 10 machines.

My add-in only works on the machine where I've build it.
On two other Windows 10 full-apdated and on Mac the server won't run as described above.

Any ideas? I may have missed something. Thanks!

**Comment by TCourtneyOwen** [comment]
@tobudim The problem is that the other machines don't have the application secret stored in the Windows Credential Manager (in the case of Windows) or the KeyChain (in the case of Mac).  At the time you ran "configure-sso" an application secret is created and added the Credential Manager or KeyChain - this secret is available in the console log after running configure-sso.  On your other machines, you would need to create a new entry that matches the name of your add-in project (i.e. the name you specified for your project when creating it via Yo Office) and use the application secret from the console log for this new entry.  You can do this in PowerShell with the script:

$ssoAppName = '<add-in project name>'
$user = '<your user name>'
$secret = '<secret>.'
[void][Windows.Security.Credentials.PasswordVault, Windows.Security.Credentials, ContentType = WindowsRuntime]
$creds = New-Object Windows.Security.Credentials.PasswordCredential
$creds.Resource = $ssoAppName
$creds.UserName = $user
$creds.Password = $secret
$vault = New-Object Windows.Security.Credentials.PasswordVault
$vault.Add($creds)

I think we could make this a bit easier for users by adding a new "Secret" param to the .ENV file that would allow users to add the secret to this file and then the office-addin-sso package could read this value if it's set.  This would make it easier to port the project over to other machines, although having a secret hard-coded into the project is not ideal (which is why we didn't take this approach initially).

I will work on making those changes today and can include you in the PR if you'd like.

Thanks,

Courtney

**Comment by tobudim** [comment]
I'll look into this tomorrow, thank you!
I'd be glad to be included in your PR. 🙂

**Comment by TCourtneyOwen** [comment]
@tobudim The PR is at https://github.com/OfficeDev/Office-Addin-Scripts/pull/346. It will also require a corresponding change to to Office-Addin-SSO to add CLIENT_SECRET to the .ENV file

**Comment by tobudim** [comment]
Thanks @TCourtneyOwen , you solved my problem.

My add-in works on both Windows and Mac now.

I do appreciate your help, have a great day.

**Comment by tobudim** [comment]
Hi @TCourtneyOwen ,

Can you confirm there is no easy solution to work on one add-in as a team? 

Setting the secret key locally (Windows and Mac) does not work.

- On Mac, the server runs but when the panel opens a 13004 error is thrown (which on the PC where I build the add-in everything's fine).
- On Windows (an other one from which I've build the add-in) I can't confirm my machine is not faulty so I'm waiting for a teammate to try it out.

Running `npm run configure-sso` on each machine is not a good idea as it creates a new app on each Azure accounts, with their own configurations.

Your PR would have been a great solution, why closing it?

I must miss something obvious as a lot of teams may work with add-ins.

I'd appreciate once again your help. 🙏

**Comment by TCourtneyOwen** [comment]
@tobudim To work on the add-in as a team, I would think that each person would need to set the secret locally in Credential Manager for Windows and the Keychain for Mac. For Windows, other team members can run the following PowerShell command  to add the secret locally:

$ssoAppName = <add-in project name>
$user = <user-name>
$secret = <secret>
[void][Windows.Security.Credentials.PasswordVault, Windows.Security.Credentials, ContentType = WindowsRuntime]
$creds = New-Object Windows.Security.Credentials.PasswordCredential
$creds.Resource = $ssoAppName
$creds.UserName = $user
$creds.Password = $secret
$vault = New-Object Windows.Security.Credentials.PasswordVault
$vault.Add($creds)

For Mac, users can run the following command from the Terminal to add the secret:

sudo security add-generic-password -a <user-name> -s <add-in- project name> -w <secret>

 I also you will need to add each team member as an owner of the application through the Azure Portal.:

![image](https://user-images.githubusercontent.com/38081557/101194341-ba6dc780-3612-11eb-9170-a1d174f69459.png)

The reason I deactivated my PR is that we are entirely comfortable with the security implications of having users write their secret to the .ENV file. Users could potentially commit the secret to their project repo and then it will no longer be a "secret".  We are going to try and come up with a better, more secure approach, but this will not happen until after the Christmas holidays.

By the way, I am going to be off for a long vacation starting tomorrow and won't be back until 01/04/21, but someone else on my team will be responding to GitHub issues in my absence.

Thanks,

Courtney

**Comment by tobudim** [comment]
Yes I did that.

Works on several Windows.
Do not work on Mac.

On Mac:
Add-in server runs.
When opening a panel, data appears.
When trying to fetch data from a server: 13004 error is thrown and add-in stop there.
Reminder: works just fine on Windows machines so the manifest must be OK.

**Comment by TCourtneyOwen** [comment]
I was able to easily port over my SSO project from Windows to Mac by creating a new password in the Mac Keychain using the same application secret I used on Windows. The new password in the Keychain must be created for the current login account, have the same name as the add-in project as specified in the manifest and use the application secret for the password. 

Also, just an FYI, I’m off for the holidays until the new year so I likely won’t respond until then, but someone else on my team is available to respond to issues. 

-Courtney

**Comment by tobudim** [comment]
Yep, since I did a lot of work while trying to deploy my add-in I've figured it out.
The problem was indeed the secret key management, as I don't want to create a new pair of key for each machine my team is working on (a new pair of key = a new app registered on Azure with it's own config).

I had to mirror office-addin-sso package in my project and put there my secret key.
I had to understand how it all works and update the scripts to launch the server.
Now I don't need to use the Keychain or Credential Manager.

For sure if I have to I won't work with add-ins for pleasure, so many troubles!

Anyway thanks for your help @TCourtneyOwen , have a great holiday. 👍
